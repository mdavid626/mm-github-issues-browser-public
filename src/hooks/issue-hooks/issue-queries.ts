import { gql } from '@apollo/client';

export const GetIssuesQuery = gql`
  query getIssues($cursor: String, $search: String!) {
    search(query: $search, after: $cursor, type: ISSUE, first: 10) {
      issueCount
      nodes {
        ... on Issue {
          id
          createdAt
          title
          state
          number
          url
          author {
            login
          }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;

export const GetIssueQuery = gql`
  query getIssue($issueNumber: Int!) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        body
        author {
          login
        }
        comments(first: 3, orderBy: { field: UPDATED_AT, direction: DESC }) {
          nodes {
            id
            createdAt
            body
            author {
              login
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

export const GetIssueCommentsQuery = gql`
  query getIssueComments($issueNumber: Int!, $commentsCursor: String) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        comments(
          first: 3
          after: $commentsCursor
          orderBy: { field: UPDATED_AT, direction: DESC }
        ) {
          nodes {
            id
            createdAt
            body
            author {
              login
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;
