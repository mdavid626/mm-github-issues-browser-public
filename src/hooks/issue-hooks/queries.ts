import { gql } from '@apollo/client';

export const GetIssuesQuery = gql`
  query getIssues($cursor: String, $states: [IssueState!]) {
    repository(owner: "facebook", name: "react") {
      id
      issues(
        first: 10
        after: $cursor
        states: $states
        orderBy: { field: CREATED_AT, direction: DESC }
      ) {
        nodes {
          id
          createdAt
          title
          state
          number
          url
          author {
            login
          }
        }
        pageInfo {
          hasNextPage
          startCursor
          endCursor
        }
        totalCount
      }
    }
  }
`;

export const GetIssueQuery = gql`
  query getIssue($issueNumber: Int!) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        bodyText
        author {
          login
        }
        comments(first: 3, orderBy: { field: UPDATED_AT, direction: DESC }) {
          nodes {
            id
            createdAt
            bodyText
            author {
              login
            }
          }
          pageInfo {
            hasPreviousPage
            hasNextPage
            startCursor
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

export const GetCommentsQuery = gql`
  query getIssueComments($issueNumber: Int!, $commentsCursor: String) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        comments(
          first: 3
          after: $commentsCursor
          orderBy: { field: UPDATED_AT, direction: DESC }
        ) {
          nodes {
            id
            createdAt
            bodyText
            author {
              login
            }
          }
          pageInfo {
            hasPreviousPage
            hasNextPage
            startCursor
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;
