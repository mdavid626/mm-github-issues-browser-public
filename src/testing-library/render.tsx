/* istanbul ignore file */
import { InitialEntry, Location } from '@remix-run/router';
import { Queries, queries } from '@testing-library/dom';
import {
  RenderHookOptions,
  RenderOptions,
  render,
  renderHook,
  RenderResult,
  RenderHookResult,
} from '@testing-library/react';
import React, { ReactElement } from 'react';
import { MemoryRouter, useLocation, useNavigate } from 'react-router-dom';
import { NavigateFunction } from 'react-router/dist/lib/hooks';

type SimpleRouter = {
  location?: Location;
  navigate?: NavigateFunction;
};

type RouterResult = {
  router: SimpleRouter;
};

export const renderWithRouter = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialEntries?: InitialEntry[]
): RenderResult & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  } as RenderResult & RouterResult;
};

export const renderHookWithRouter = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialEntries?: InitialEntry[]
): RenderHookResult<Result, Props> & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  };
};
