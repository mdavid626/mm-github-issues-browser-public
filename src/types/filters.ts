export const stateFilters = ['open', 'closed'] as const;

export type StateFilter = (typeof stateFilters)[number];

export type Filters = {
  search: string;
  state: StateFilter | null;
};
