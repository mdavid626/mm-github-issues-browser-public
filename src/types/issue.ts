import { PageInfo } from './pagination';

export type IssueState = 'OPEN' | 'CLOSED';

export type IssuesQueryItem = {
  id: string;
  createdAt: string;
  title: string;
  state: IssueState;
  number: number;
  url: string;
  author: {
    login: string;
  };
};

export type IssuesQueryResult = {
  repository: {
    issues: {
      nodes: IssuesQueryItem[];
      pageInfo: PageInfo;
      totalCount: number;
    };
  };
};

export type IssueComment = {
  id: string;
  bodyText: string;
  author: {
    login: string;
  };
  createdAt: string;
};

export type IssueQueryItem = {
  id: string;
  createdAt: string;
  title: string;
  state: IssueState;
  number: number;
  url: string;
  bodyText: string;
  author: {
    login: string;
  };
  comments: {
    nodes: IssueComment[];
    pageInfo: PageInfo;
    totalCount: number;
  };
};

export type IssueQueryResult = {
  repository: {
    issue: IssueQueryItem;
  };
};
