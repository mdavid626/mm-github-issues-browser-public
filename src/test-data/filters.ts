import { Filters } from '../types/filters';

export const testFilters1: Filters = {
  search: 'test',
  state: 'open',
};
