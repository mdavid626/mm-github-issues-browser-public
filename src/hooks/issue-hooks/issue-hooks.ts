import { useQuery, gql } from '@apollo/client';
import { ApolloError } from '@apollo/client/errors';
import { IssueQueryResult, IssuesQueryResult } from '../../types/issue';

const issuesQuery = gql`
  query {
    repository(owner: "facebook", name: "react") {
      issues(last: 20) {
        edges {
          node {
            id
            createdAt
            title
            state
            number
            url
            author {
              login
            }
          }
        }
      }
    }
  }
`;

export const useIssues = (): [
  IssuesQueryResult | undefined,
  boolean,
  ApolloError | undefined
] => {
  const { data, loading, error } = useQuery<IssuesQueryResult>(issuesQuery);
  return [data, loading, error];
};

const issueQuery = gql`
  {
    repository(owner: "facebook", name: "react") {
      issue(number: 26077) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        bodyText
        author {
          login
        }
      }
    }
  }
`;

export const useIssue = (): [
  IssueQueryResult | undefined,
  boolean,
  ApolloError | undefined
] => {
  const { data, loading, error } = useQuery<IssueQueryResult>(issueQuery);
  return [data, loading, error];
};
