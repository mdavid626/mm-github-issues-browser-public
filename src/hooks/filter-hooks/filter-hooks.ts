import { useCallback, useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { Filters, StateFilter, stateFilters } from '../../types/filters';

export const useFilters = (): [Filters, (newFilters: Filters) => void] => {
  const location = useLocation();
  const navigate = useNavigate();

  const filters = useMemo(() => {
    const query = new URLSearchParams(location.search);
    const state = query.get('state') as StateFilter;
    return {
      search: query.get('search') ?? '',
      state: stateFilters.includes(state) ? state : null,
    };
  }, [location.search]);

  const setFilters = useCallback(
    (newFilters: Filters) => {
      const query = new URLSearchParams(location.search);
      query.set('search', newFilters.search);
      if (newFilters.state) {
        query.set('state', newFilters.state);
      } else {
        query.delete('state');
      }
      navigate({
        search: query.toString(),
      });
    },
    [location.search, navigate]
  );

  return [filters, setFilters];
};
