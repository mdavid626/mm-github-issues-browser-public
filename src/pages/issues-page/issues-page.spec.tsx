import { cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import FiltersBar from '../../components/filters-bar/filters-bar';
import Footer from '../../components/footer/footer';
import Header from '../../components/header/header';
import IssueListItem from '../../components/issue-list-item/issue-list-item';
import { useFilters } from '../../hooks/filter-hooks/filter-hooks';
import { useIssues } from '../../hooks/issue-hooks/issue-hooks';
import { testFilters1 } from '../../test-data/filters';
import {
  testEmptyIssuesQueryResult1,
  testIssuesQueryResult1,
} from '../../test-data/issues';
import { renderWithRouter } from '../../testing-library/render';
import { IssuesItem } from '../../types/issue';
import IssuesPage from './issues-page';

jest.mock('../../components/filters-bar/filters-bar');
jest.mock('../../components/footer/footer');
jest.mock('../../components/header/header');
jest.mock('../../components/issue-list-item/issue-list-item');
jest.mock('../../hooks/filter-hooks/filter-hooks');
jest.mock('../../hooks/issue-hooks/issue-hooks');

describe('issues-page', () => {
  beforeEach(() => {
    (Footer as jest.Mock).mockReturnValue(<div>footer</div>);
    (Header as jest.Mock).mockReturnValue(<div>header</div>);
    (FiltersBar as jest.Mock).mockReturnValue(<div>filters-bar</div>);
    (IssueListItem as jest.Mock).mockImplementation(
      ({ issue }: { issue: IssuesItem }) => (
        <div>issue-list-item-{issue.number}</div>
      )
    );
    (useFilters as jest.Mock).mockReturnValue([testFilters1, jest.fn()]);
    (useIssues as jest.Mock).mockReturnValue([
      testIssuesQueryResult1,
      false,
      undefined,
      jest.fn(),
    ]);
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  it('should render', () => {
    const setFilters = jest.fn();
    (useFilters as jest.Mock).mockReturnValue([testFilters1, setFilters]);
    const { asFragment } = renderWithRouter(<IssuesPage />);
    expect(asFragment()).toMatchSnapshot();
    expect(useFilters).toHaveBeenCalledWith();
    expect(useIssues).toHaveBeenCalledWith(testFilters1);
    expect(Footer).toHaveBeenCalledWith({}, {});
    expect(Header).toHaveBeenCalledWith({}, {});
    expect(FiltersBar).toHaveBeenCalledWith(
      { filters: testFilters1, setFilters },
      {}
    );
    expect(IssueListItem).toHaveBeenCalledTimes(2);
    expect(IssueListItem).toHaveBeenCalledWith(
      { issue: testIssuesQueryResult1.search.nodes[0] },
      {}
    );
    expect(IssueListItem).toHaveBeenCalledWith(
      { issue: testIssuesQueryResult1.search.nodes[1] },
      {}
    );
  });

  it('should render spinner when loading', () => {
    (useIssues as jest.Mock).mockReturnValue([
      undefined,
      false,
      undefined,
      jest.fn(),
    ]);
    const { asFragment } = renderWithRouter(<IssuesPage />);
    expect(asFragment()).toMatchSnapshot();
    expect(Footer).toHaveBeenCalledWith({}, {});
    expect(Header).toHaveBeenCalledWith({}, {});
    expect(FiltersBar).toHaveBeenCalledWith(
      { filters: testFilters1, setFilters: expect.any(Function) },
      {}
    );
    expect(IssueListItem).not.toHaveBeenCalled();
  });

  it('should show error when error', () => {
    (useIssues as jest.Mock).mockReturnValue([
      undefined,
      false,
      new Error('test error'),
      jest.fn(),
    ]);
    renderWithRouter(<IssuesPage />);
    expect(screen.getByText('test error')).toBeVisible();
    expect(Footer).toHaveBeenCalledWith({}, {});
    expect(Header).toHaveBeenCalledWith({}, {});
    expect(FiltersBar).toHaveBeenCalledWith(
      { filters: testFilters1, setFilters: expect.any(Function) },
      {}
    );
    expect(IssueListItem).not.toHaveBeenCalled();
  });

  it('should go to issue detail page when issue clicked', async () => {
    const { router } = renderWithRouter(<IssuesPage />, undefined, [
      '/?test=1',
    ]);
    await userEvent.click(screen.getByText('issue-list-item-26126'));
    expect(router.location?.pathname).toBe('/issue/26126');
    expect(router.location?.search).toBe('?test=1');
  });

  it('should load more items when load more clicked', async () => {
    const fetchMore = jest.fn();
    (useIssues as jest.Mock).mockReturnValue([
      testIssuesQueryResult1,
      false,
      undefined,
      fetchMore,
    ]);
    renderWithRouter(<IssuesPage />);
    await userEvent.click(screen.getByText('load more'));
    expect(fetchMore).toHaveBeenCalledWith();
  });

  it('should not show load more when no more pages', async () => {
    (useIssues as jest.Mock).mockReturnValue([
      testEmptyIssuesQueryResult1,
      false,
      undefined,
      jest.fn(),
    ]);
    renderWithRouter(<IssuesPage />);
    expect(screen.queryByText('load more')).toBe(null);
  });

  it('should show loading indicator when loading more', async () => {
    (useIssues as jest.Mock).mockReturnValue([
      testIssuesQueryResult1,
      true,
      undefined,
      jest.fn(),
    ]);
    renderWithRouter(<IssuesPage />);
    expect(screen.getByText('loading...')).toBeVisible();
    expect(screen.queryByText('load more')).toBe(null);
  });
});
