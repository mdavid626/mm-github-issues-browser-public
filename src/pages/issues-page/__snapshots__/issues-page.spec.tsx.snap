// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issues-page should render 1`] = `
<DocumentFragment>
  <div
    class="IssuesPage"
  >
    <div>
      header
    </div>
    <div
      class="IssuesPage-content"
    >
      <div
        class="IssuesPage-filters"
      >
        <div>
          filters-bar
        </div>
        <div
          class="IssuesPage-totalCount"
        >
          Total Issues: 2
        </div>
      </div>
      <div
        class="IssuesPage-issues"
      >
        <a
          class="IssuesPage-issue"
          href="/issue/26126"
        >
          <div>
            issue-list-item-26126
          </div>
        </a>
        <a
          class="IssuesPage-issue"
          href="/issue/26130"
        >
          <div>
            issue-list-item-26130
          </div>
        </a>
      </div>
      <div
        class="IssuesPage-loadMore"
      >
        load more
      </div>
    </div>
    <div>
      footer
    </div>
  </div>
</DocumentFragment>
`;

exports[`issues-page should render spinner when loading 1`] = `
<DocumentFragment>
  <div
    class="IssuesPage"
  >
    <div>
      header
    </div>
    <div
      class="IssuesPage-content"
    >
      <div
        class="IssuesPage-filters"
      >
        <div>
          filters-bar
        </div>
      </div>
      <img
        alt="loading spinner"
        class="PageLoader-spinner"
        data-testid="PageLoader-spinner"
        src="spinner.svg"
      />
    </div>
    <div>
      footer
    </div>
  </div>
</DocumentFragment>
`;
